Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+        return m_displayLayer.get();
+#if HAVE(AVSAMPLEBUFFERVIDEORENDERER)
+    return m_renderer.get();
+#else
+    return nil;
+#endif
 }
 
 AVSampleBufferDisplayLayer *VideoMediaSampleRenderer::displayLayer() const
 {
-    return dynamic_objc_cast<AVSampleBufferDisplayLayer>(m_renderer.get(), PAL::getAVSampleBufferDisplayLayerClass());
+    assertIsMainThread();
+    return m_displayLayer.get();
 }
 
+RetainPtr<CVPixelBufferRef> VideoMediaSampleRenderer::copyDisplayedPixelBuffer()
+{
+    assertIsMainThread();
+
 #if HAVE(AVSAMPLEBUFFERDISPLAYLAYER_COPYDISPLAYEDPIXELBUFFER)
+    if (auto buffer = adoptCF([renderer() copyDisplayedPixelBuffer]))
+        return buffer;
+#endif
 
-RetainPtr<CVPixelBufferRef> VideoMediaSampleRenderer::copyDisplayedPixelBuffer() const
-{
-    return adoptCF([m_renderer copyDisplayedPixelBuffer]);
+    RetainPtr<CVPixelBufferRef> imageBuffer;
+
+    m_workQueue->dispatchSync([&] {
+        if (!m_timebase)
+            return;
+
+        purgeDecodedSampleQueue();
+
+        CMTime currentTime = PAL::CMTimebaseGetTime(m_timebase.get());
+        auto nextSample = (CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueGetHead(m_decodedSampleQueue.get()));
+        CMTime presentationTime = PAL::CMSampleBufferGetOutputPresentationTimeStamp(nextSample);
+
+        if (PAL::CMTimeCompare(presentationTime, currentTime) > 0)
+            return;
+
+        RetainPtr sampleToBePurged = adoptCF((CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueDequeueAndRetain(m_decodedSampleQueue.get())));
+        imageBuffer = (CVPixelBufferRef)PAL::CMSampleBufferGetImageBuffer(sampleToBePurged.get());
+    });
+
+    if (!imageBuffer)
+        return nullptr;
+
+    ASSERT(CFGetTypeID(imageBuffer.get()) == CVPixelBufferGetTypeID());
+    if (CFGetTypeID(imageBuffer.get()) != CVPixelBufferGetTypeID())
+        return nullptr;
+    return imageBuffer;
 }
 
 CGRect VideoMediaSampleRenderer::bounds() const
 {
     return [displayLayer() bounds];
 }
 
-#endif
+unsigned VideoMediaSampleRenderer::totalVideoFrames() const
+{
+    if (m_decompressionSession)
+        return m_decompressionSession->totalVideoFrames();
+    return [renderer() videoPerformanceMetrics].totalNumberOfVideoFrames;
+}
+
+unsigned VideoMediaSampleRenderer::droppedVideoFrames() const
+{
+    if (m_decompressionSession)
+        return m_decompressionSession->droppedVideoFrames();
+    return [renderer() videoPerformanceMetrics].numberOfDroppedVideoFrames;
+}
+
+unsigned VideoMediaSampleRenderer::corruptedVideoFrames() const
+{
+    if (m_decompressionSession)
+        return m_decompressionSession->corruptedVideoFrames();
+    return [renderer() videoPerformanceMetrics].numberOfCorruptedVideoFrames;
+}
+
+MediaTime VideoMediaSampleRenderer::totalFrameDelay() const
+{
+    if (m_decompressionSession)
+        return m_decompressionSession->totalFrameDelay();
+    return MediaTime::createWithDouble([renderer() videoPerformanceMetrics].totalFrameDelay);
```

Comment:
should we add the two values if we have a decompression ses

Reply:
You can't delay a decompressed video frame either.